File: rest_api/src/circuitBreaker.js
const httpClient = require("./httpClient.js")();

class CircuitBreaker {
    // Open, Half, Closed
    CircuitBreakerState;
    successCount;
    failCount;

    // Timeout für den Reset von successCount und failCount im Status CLOSED
    timeoutReset;

    // Timeout für die Circuit Breaker soll bei 30 Sekunden liegen
    timeoutOpenState;

    // Verhältnis successCount und failCount:
    // Verhältnis wann Half State ausgelöst werden soll
    triggerHalfState;
    // Verhältnis wann Open State ausgelöst werden soll
    triggerOpenState;
    // Verhältnis wann Closed State ausgelöst werden soll
    triggerClosedState;

    timestamp;

    // Nur benötigt im Zustand HALF
    permittedRequestsInStateHalf;
    requestCount;

    // Circuit Breaker wird für jeden Host einzeln erstellt
    hostname;
    port;

    constructor(timeoutReset, timeoutOpenState, triggerHalfState,
                triggerOpenState, permittedRequestsInStateHalf,
                triggerClosedState, hostname, port) {
        this.failCount = 0;
        this.successCount = 0;
        this.requestCount = 0;
        this.timeoutReset = timeoutReset;
        this.timeoutOpenState = timeoutOpenState;
        this.triggerHalfState = triggerHalfState;
        this.triggerOpenState = triggerOpenState;
        this.triggerClosedState = triggerClosedState;
        this.permittedRequestsInStateHalf = permittedRequestsInStateHalf;
        this.CircuitBreakerState = "CLOSED";
        this.timestamp = new Date();
        this.hostname = hostname;
        this.port = port;
    }


    checkReset(timeDiff) {
        console.log(timeDiff);
        console.log(timeDiff - this.timeoutReset);
        console.log("Circuit Breaker: Prüfe ob Circuit Breaker Status zurückgesetzt werden soll");
        if(timeDiff > this.timeoutReset &&
            (this.CircuitBreakerState == "CLOSED" || this.CircuitBreakerState == "HALF")) {
            console.log("Circuit Breaker: Kompletter Status wird zurückgesetzt!");
            this.failCount = 0;
            this.successCount = 0;
            this.timestamp = new Date();
            this.requestCount = 0;
        }
    }


    // hostname = Loadbalancer der an die entsprechenden Microservices innerhalb der Fachlichkeit weiterleiten
    async circuitBreakerRequest(path, login_name, auth_token, httpMethod) {

        // Schritt 1: Berechne Abstand zwischen gespeicherten timeStamp und aktuellen timeStamp in Sekunden
        let timeDiff = ( new Date() - this.timestamp ) / 1000;

        // Schritt 2: Wenn der Timestamp älter ist als 5 Minuten -> setze alles auf Anfang
        this.checkReset(timeDiff);

        //Schritt 3: Prüfe ob timeout für Zustand Open abgelaufen ist
        if(this.CircuitBreakerState == "OPEN") {

            if(timeDiff >= this.timeoutOpenState) {
                // Wenn timeout abgelaufen setze den Circuit Breaker wieder auf HALF
                console.log("Circuit Breaker: Wechsel Circuit Breaker Status von OPEN auf HALF");
                this.CircuitBreakerState = "HALF";

            } else {
                console.log("Circuit Breaker: immer noch auf Zustand OPEN");
                console.log("Circuit Breaker ist "+ (this.timeoutOpenState - timeDiff)  + " noch offen");
                throw "Request fehlgeschlagen: Circuit Breaker ist im Zustand offen. Keine Requests zu Service Benutzerverwaltung erlaubt";
            }

        }

        // Schritt 4: Prüfe ob Circuit Breaker im Zustand HALF ist -> Wenn Ja dann prüfe ob das Limit an Requests erreicht ist
        if(this.CircuitBreakerState == "HALF" && this.requestCount > this.permittedRequestsInStateHalf) {
            console.log("Request fehlgeschlagen: Circuit Breaker ist auf Zustand HALF aber der erlaubte RequestCount ist erreicht")
            throw "Request fehlgeschlagen: Circuit Breaker ist auf Zustand HALF aber der erlaubte RequestCount ist erreicht";
        }

        // Schritt 5: Prüfe ob wieder auf Closed gesetzt werden kann
        if(this.CircuitBreakerState == "HALF" && (this.successCount - this.failCount > this.triggerClosedState)) {
            this.checkReset(timeDiff)
            console.log("Circuit Breaker: Wechsel Circuit Breaker Status von HALF auf CLOSED")
            this.CircuitBreakerState = "CLOSED";
        }

        let headerData = {
            'auth_token': auth_token,
            'login_name': login_name
        };

        try {
            // Schritt 6: Falls die Bedingungen von Schritt 1 - Schritt 4 nicht erfüllt sind -> führe den Request durch
            console.log("Circuit Breaker: Führe HTTP Request im Circuit Breaker durch");
            // Wenn auf HALF gesetzt dann zähle die Anzahl der Requests mit
            // Deshalb so gelöst, da requestCount nur im Zustand HALF benötigt wird
            if(this.CircuitBreakerState == "HALF") {
                this.requestCount++;
            }
            let result = await httpClient.makeRequest(this.hostname, this.port, path, {}, headerData, httpMethod);
            this.successCount++;
            console.log("Circuit Breaker: Request war erfolgreich. Success Count ist jetzt bei " + this.successCount);
            return JSON.stringify(result);
        } catch (err) {
            this.failCount++;

            // Schritt 7: Prüfe das Verhältnis zwischen successCount und failCount
            // Schritt 7b: Prüfe ob auf HALF gesetzt werden muss
            if(this.CircuitBreakerState == "CLOSED" && this.successCount - this.failCount < this.triggerHalfState) {
                console.log("Circuit Breaker: Wechsel Circuit Breaker Status von CLOSED auf HALF");
                this.CircuitBreakerState = "HALF";
                this.timestamp = new Date();
            }

            // Schritt 7c: Prüfe ob auf Open gesetzt werden muss
            if(this.CircuitBreakerState == "HALF" && (this.successCount - this.failCount < this.triggerOpenState)) {
                console.log("Circuit Breaker: Wechsel Circuit Breaker Status von HALF auf OPEN");
                this.CircuitBreakerState = "OPEN";
                this.timestamp = new Date();
            }
            console.log("Circuit Breaker: Request ist fehlgeschlagen. Fail Count ist jetzt bei " + this.failCount);
            throw err;
        }
    }


}

module.exports = CircuitBreaker

File: rest_api/src/circuitbreaker.rs
use chrono::{DateTime, Utc};
use anyhow::anyhow;
#[path = "./client.rs"] mod client;
use std::sync::{Arc, Mutex};

#[derive(Clone)]
pub struct CircuitBreaker<'a> {
    circuit_breaker_state: Arc<std::sync::Mutex<&'a str>>,
    success_count: Arc<std::sync::Mutex<i64>>,
    fail_count: Arc<std::sync::Mutex<i64>>,
    timeout_reset: i64,
    timeout_open_state: i64,
    trigger_half_state: i64,
    trigger_open_state: i64,
    trigger_closed_state: i64,
    timestamp:  Arc<std::sync::Mutex<DateTime<Utc>>>,
    permitted_requests_in_state_half: i64,
    request_count: Arc<std::sync::Mutex<i64>>,
    hostname:  &'a str,
    port: i32,
}
impl <'a> CircuitBreaker<'a>  {
    pub fn new(timeout_reset: i64, timeout_open_state: i64, trigger_half_state: i64, trigger_open_state: i64,
               permitted_requests_in_state_half: i64, trigger_closed_state: i64, hostname: &'a str, port: i32) -> Self {

        return Self { circuit_breaker_state: Arc::new(Mutex::new("CLOSED")), success_count: Arc::new(Mutex::new(0)), fail_count: Arc::new(Mutex::new(0)), timeout_reset, timeout_open_state,
            trigger_half_state, trigger_closed_state, timestamp: Arc::new(Mutex::new(Utc::now())), trigger_open_state, permitted_requests_in_state_half,
            request_count: Arc::new(Mutex::new(0)), hostname, port};
    }

    fn check_reset(&mut self, time_diff: i64) {
        println!("{}", time_diff);
        println!("{}", time_diff - self.timeout_reset);
        println!("Circuit Breaker: Prüfe ob Circuit Breaker Status zurückgesetzt werden soll");
        if time_diff > self.timeout_reset &&
            (self.get_circuit_breaker_state() == "CLOSED" || self.get_circuit_breaker_state() == "HALF") {
            println!("Circuit Breaker: Kompletter Status wird zurückgesetzt!");
            *(self.fail_count.lock().unwrap()) = 0;
            *(self.success_count.lock().unwrap()) = 0;
            *(self.timestamp.lock().unwrap()) = Utc::now();
            *(self.request_count.lock().unwrap()) = 0;
        }
    }

    pub async fn circuit_breaker_post_request(&mut self, addr_with_params: String, login_name: String, auth_token: String, http_method: String) -> Result<(http_req::response::Response, String), anyhow::Error> {

        println!("REST API: AKTUELLER CIRCUIT BREAKER STATUS IST: {}", self.get_circuit_breaker_state());

        let current_timestamp = Utc::now();
        let time_diff = current_timestamp.signed_duration_since(self.get_timestamp()).num_seconds();
        println!("timeDiff is {}", time_diff);
        println!("timeout_open_state ist {}", self.timeout_open_state);

        self.check_reset(time_diff);

        if self.get_circuit_breaker_state() == "OPEN" {

            if time_diff >= self.timeout_open_state {
                // Wenn timeout abgelaufen setze den Circuit Breaker wieder auf HALF
                println!("Circuit Breaker: Wechsel Circuit Breaker Status von OPEN auf HALF");
                self.set_circuit_breaker_state("HALF");

            } else {
                println!("Circuit Breaker: immer noch auf Zustand OPEN");
                println!("Circuit Breaker ist {} noch offen", (self.timeout_open_state - time_diff));
                return Err(anyhow!("Request fehlgeschlagen: Circuit Breaker ist im Zustand offen und keine Requests sind zum Service Benutzerverwaltung erlaubt"))
            }

        }

        if self.get_circuit_breaker_state() == "HALF" && self.get_request_count() > self.permitted_requests_in_state_half {
            println!("Request fehlgeschlagen: Circuit Breaker ist auf Zustand HALF aber der erlaubte RequestCount ist erreicht");
            return Err(anyhow!("Request fehlgeschlagen: Circuit Breaker ist auf Zustand HALF aber der erlaubte RequestCount ist erreicht"))
        }

        if self.get_circuit_breaker_state() == "HALF" && (self.get_success_count() - self.get_fail_count() > self.trigger_closed_state) {
            self.check_reset(time_diff);
            println!("Circuit Breaker: Wechsel Circuit Breaker Status von HALF auf CLOSED");
            self.set_circuit_breaker_state("CLOSED");
        }

        println!("Circuit Breaker: Führe HTTP Request im Circuit Breaker durch");
        if self.get_circuit_breaker_state() == "HALF" {
            self.increment_request_count();
        }

        match client::make_request(self.hostname, self.port, addr_with_params, login_name, auth_token, http_method).await {

            Ok((res, response_json_string)) => {
                self.increment_success_count();
                println!("Circuit Breaker: Request war erfolgreich. Success Count ist jetzt bei {}", self.get_success_count());

                if res.status_code().is_success() {
                    return Ok((res, response_json_string))
                } else {
                    return Err(anyhow!("CircuitBreaker: Request failed, return code was {}", res.status_code()))
                }

            },
            Err(err) => {
                println!("{:?}", err);

                self.increment_fail_count();
                if self.get_circuit_breaker_state() == "CLOSED" && self.get_success_count() - self.get_fail_count() < self.trigger_half_state {
                    println!("Circuit Breaker: Wechsel Circuit Breaker Status von CLOSED auf HALF");
                    self.set_circuit_breaker_state("HALF");
                    *(self.timestamp.lock().unwrap()) = Utc::now();
                }

                if self.get_circuit_breaker_state() == "HALF" && (self.get_success_count() - self.get_fail_count() < self.trigger_open_state) {
                    println!("Circuit Breaker: Wechsel Circuit Breaker Status von HALF auf OPEN");
                    self.set_circuit_breaker_state("OPEN");
                    *(self.timestamp.lock().unwrap()) = Utc::now();
                }

                println!("Circuit Breaker: Request ist fehlgeschlagen. Fail Count ist jetzt bei {}", self.get_fail_count());
                return Err(anyhow!("CircuitBreaker: Request failed, return code was 500"))
            },
        }

    }

    // Mutex Variablen: Jeweils der Zugriff als auch das Schreiben wird mit Mutex Sperre belegt
    // Nach dem Block wird jeweils immer die Variable freigegeben
    // Deshalb auch hier die Read operationen in dem Block
    // Wenn sich während der Prüfung der Variablen sich etwas ändert ist das nicht schlimm
    // Beim Schreiben muss es aber sicher sein

    fn increment_success_count(&self) {
        let mut counter = self.success_count.lock().unwrap();
        *counter += 1;
        println!("REST API: Success Count ist {}", *counter);
    }

    fn increment_fail_count(&self) {
        let mut counter = self.fail_count.lock().unwrap();
        *counter += 1;
        println!("REST API: Fail Count ist {}", *counter);
    }

    fn increment_request_count(&self) {
        let mut counter = self.request_count.lock().unwrap();
        *counter += 1;
        println!("REST API: Request Count ist {}", *counter);
    }

    fn set_circuit_breaker_state(&self, new_state: &'a str) {
        let mut state = self.circuit_breaker_state.lock().unwrap();
        *state = new_state;
    }

    fn get_request_count(&self) -> i64 {
        let counter = self.request_count.lock().unwrap();
        return *counter;
    }

    fn get_success_count(&self) -> i64 {
        let counter = self.success_count.lock().unwrap();
        return *counter;
    }

    fn get_fail_count(&self) -> i64{
        let counter = self.fail_count.lock().unwrap();
        return *counter;
    }

    fn get_timestamp(&self) -> DateTime<Utc> {
        let timestamp = self.timestamp.lock().unwrap();
        return *timestamp;
    }

    fn get_circuit_breaker_state(&self) -> &str {
        let state = self.circuit_breaker_state.lock().unwrap();
        return *state;
    }



}


